jmix-reports/reports/src/main/java/io/jmix/reports/impl/repository/ReportRepositoryImpl.java

```diff
@@ -0,0 +1,388 @@
+        }
+        return true;
+    }
+
+    protected boolean outputTypeSatisfies(Report report, @Nullable ReportOutputType outputType) {
+        if (report.getSource() == ReportSource.DATABASE) {
+            // we rely on this condition being checked in ReportSecurityManager, on data loading stage.
+            // moreover, templates aren't loaded from db here
+            return true;
+        }
+        if (outputType != null) {
+            return report.getTemplates().stream()
+                    .anyMatch(t -> t.getReportOutputType() == outputType);
+        }
+        return true;
+    }
+
+    @Override
+    public int getTotalCount(ReportFilter filter) {
+        if (!isReadPermitted()) {
+            return 0;
+        }
+        Collection<Report> annotatedReports = annotatedReportHolder.getAllReports();
+        List<Report> reportsFromDb = loadReportsFromDatabase(filter);
+
+        FilteringContext filteringContext = fillFilteringContext(filter);
+        long count = Stream.concat(annotatedReports.stream(), reportsFromDb.stream())
+                .filter(rg -> satisfies(rg, filter, filteringContext))
+                .count();
+
+        return (int) count;
+    }
+
+    protected List<Report> loadReportsFromDatabase(ReportFilter filter) {
+        // some of filtering conditions are applied right now, at the moment of loading from DB
+        //  later they are skipped during in-memory filtering phase
+        List<Report> reportsFromDb = reportSecurityManager.getAvailableReports(
+                filter.getViewId(),
+                filter.getUser(),
+                filter.getInputValueMetaClass(),
+                filter.getSystem(),
+                filter.getOutputType(),
+                null
+        );
+        return reportsFromDb;
+    }
+
+    @Nullable
+    @Override
+    public Report loadForRunningByCode(String reportCode) {
+        if (!isReadPermitted()) {
+            return null;
+        }
+        Report report = annotatedReportHolder.getByCode(reportCode);
+        if (report != null) {
+            return report;
+        }
+        report = dataManager.load(Report.class)
+                .condition(PropertyCondition.equal("code", reportCode))
+                .fetchPlan(RUN_FETCH_PLAN)
+                .optional()
+                .orElse(null);
+
+        return report;
+    }
+
+    @Override
+    public boolean existsReportByCode(String reportCode) {
+        if (!isReadPermitted()) {
+            throw new AccessDeniedException("entity", metadata.getClass(Report.class).getName(), EntityOp.READ.getId());
+        }
+
+        for (Report report : annotatedReportHolder.getAllReports()) {
+            if (report.getCode() != null && report.getCode().equals(reportCode)) {
+                return true;
+            }
+        }
+
+        Optional<Report> report = dataManager.load(Report.class)
```

Comment:
Use `dataManager.loadValue("select count(e) from ...", Long.class)` - this is more efficient than loading entire entity